# zebrafinchRA
R scripts for RNA-seq analysis that assess developmental and sex-biased gene expression in zebra finch RA. Detailed methods and results are reported in [Friedrich et al., 2022](TBD).

## Getting started

The easiest way to get going is to load dependencies via the `renv` virtual environment included with this project. 
- First, if not already installed on your machine, please install the [`renv` package](https://rstudio.github.io/renv/articles/renv.html). 
- Then, when opening R scripts from the main project directory (`zebrafinchRA/`), `renv` should automatically detect and set up the virtual environment with all dependencies outlined in the lockfile (`renv.lock`).
- Once all dependencies are loaded in the virtual environment, run the `deseq_analysis.R` file.

Alternatively, manually install the [packages required](#dependencies) (listed below), then run `deseq_analysis.R`.

Note: Several functions in this project rely on the [xlsx package](https://cran.r-project.org/web/packages/xlsx/) to export R objects to Excel files. The xlsx package has Java dependencies, so Java must be installed on the user's machine for results table export functions to work.

## Files
### Analysis files
- `deseq_analysis.R` is the main analysis pipeline; it runs/loads `make_count_and_feature_tables.R` and `de_functions.R` before proceeding with analysis
- `de_functions.R` contains custom analysis and plotting functions called by `deseq_analysis.R`
- `make_count_and_feature_tables.R` generates the count table "`cts`" (main input to `DESeq2`) and other supporting tables from data supplied in `input_files/`

### Input files
All necessary input data is located within the `input_files/` directory.
- `STAR_counts/` directory contains each sample's gene count table (specifically, ReadsPerGene.out.tab files generated by [STAR](https://github.com/alexdobin/STAR))
- `GCF_003957565.1_bTaeGut1_v1.p_feature_table.csv` contains feature annotations for the zebra finch RefSeq assembly [GCF_003957565.1](https://www.ncbi.nlm.nih.gov/assembly/GCF_003957565.1/)
- `ZEBrA_MarkerTable_070721.csv` contains the RA marker status for genes with *in situ* expression patterns in the [ZEBrA Database](http://www.zebrafinchatlas.org/)
- `manual_annotations.tsv` contains a dictionary that maps gene symbols to their associated RefSeq LOC (confirmed by BLAST)
- `sample_info.csv` maps RNA-seq sample names to age and sex groups
- `wzpairs.csv` contains a list of genes that have copies on both the W and Z sex chromosomes

## Raw reads accessibility

The data discussed in this publication have been deposited in NCBI's Gene Expression Omnibus (Edgar et al., 2002) and are accessible through GEO Series accession number GSE191296 (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE191296).

## Dependencies<a name="dependencies"></a>
```
library(plyr)
library(tidyverse)
library(DESeq2)
library(xlsx)
library(DEGreport)
library(grid)
library(gridExtra)
library(annotables)
library(org.Hs.eg.db)
library(clusterProfiler)
```

## Author
Sami Friedrich, PhD
- [Linkedin](https://www.linkedin.com/in/sami-friedrich/)
